deduper.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2021 by Farsight Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "deduper.h"
+
+struct chainlink;
+typedef struct chainlink *chainlink_t;
+
+struct chainlink {
+	chainlink_t next;
+	char str[];
+};
+static inline size_t chainlink_size(size_t length) {
+	return sizeof(struct chainlink) + length + 1;
+}
+
+struct deduper {
+	size_t buckets;
+	chainlink_t chains[];
+};
```

Comment:
add newline

Reply:
this inline function would normally be a macro (##define) but since all our target compilers are modern, i used a static inline function. the logic here should be as close as possible to the structure definition, since any change to the structure will require changes to the inline function. since it is not a "normal" function, i did not use a newline to offset it. if you think it would look better with a newline, i have no objections.